File: src/utils/retry.ts
import type { Settings, RetryConfig } from '../types/config.js';

const PARSE_DURATION_REGEX = /^(\d+)(s|m|h)?$/;

export function parseDuration(duration: string): number {
  const match = duration.match(PARSE_DURATION_REGEX);
  if (!match) {
    return 0;
  }
  
  const value = parseInt(match[1], 10);
  const unit = match[2] || 's';
  
  switch (unit) {
    case 's':
      return value * 1000;
    case 'm':
      return value * 60 * 1000;
    case 'h':
      return value * 60 * 60 * 1000;
    default:
      return value * 1000;
  }
}

export function calculateBackoff(
  attempt: number,
  config: RetryConfig
): number {
  const initial = parseDuration(config.initial_delay);
  const max = parseDuration(config.max_delay);
  const factor = config.factor;
  
  const delay = initial * Math.pow(factor, attempt - 1);
  return Math.min(delay, max);
}

export function shouldRetry(
  attempt: number,
  maxAttempts: number,
  error: Error,
  retryableErrors?: string[]
): boolean {
  if (attempt >= maxAttempts) {
    return false;
  }
  
  if (!retryableErrors || retryableErrors.length === 0) {
    return true;
  }
  
  const errorMessage = error.message.toLowerCase();
  return retryableErrors.some((re) => errorMessage.includes(re.toLowerCase()));
}

export async function withRetry<T>(
  fn: () => Promise<T>,
  settings: Settings
): Promise<T> {
  const config = settings.retry || {
    max_attempts: 5,
    initial_delay: '1s',
    max_delay: '30s',
    factor: 2.0,
  };
  
  const retryableErrors = ['network', 'timeout', 'connection', 'econnreset'];
  
  let attempt = 0;
  let lastError: Error | undefined;
  
  while (attempt < config.max_attempts) {
    attempt++;
    
    try {
      return await fn();
    } catch (error) {
      lastError = error instanceof Error ? error : new Error(String(error));
      
      if (!shouldRetry(attempt, config.max_attempts, lastError, retryableErrors)) {
        throw lastError;
      }
      
      const delay = calculateBackoff(attempt, config);
      await new Promise((resolve) => setTimeout(resolve, delay));
    }
  }
  
  throw lastError;
}

File: src/git_sync/utils/retry.py
from tenacity import retry, retry_if_exception_type, stop_after_attempt, wait_exponential

from git_sync.models.config import RetryConfig, Settings


def parse_duration(duration: str) -> float:
    match = duration.match(r"^(\d+)(s|m|h)?$")
    if not match:
        return 0

    value = int(match.group(1))
    unit = match.group(2) or "s"

    multipliers = {"s": 1, "m": 60, "h": 3600}
    return value * multipliers.get(unit, 1)


def get_retry_decorator(settings: Settings) -> retry:
    config = settings.retry or RetryConfig()

    return retry(
        stop=stop_after_attempt(config.max_attempts),
        wait=wait_exponential(
            multiplier=parse_duration(config.initial_delay),
            max=parse_duration(config.max_delay),
            exp_base=config.factor,
        ),
        retry=retry_if_exception_type((ConnectionError, TimeoutError, OSError)),
    )


def with_retry(func, settings: Settings):
    decorator = get_retry_decorator(settings)
    return decorator(func)
